future/atomicTransactionComposer.go

```diff
@@ -0,0 +1,500 @@
+		txAndSigner.Txn.Group = types.Digest{}
+	}
+
+	return AtomicTransactionComposer{
+		status:       BUILDING,
+		transactions: newTxs,
+		methodCalls:  atc.methodCalls,
+		signedTxs:    [][]byte{},
+		txids:        []string{},
+	}
+}
+
+/**
+* Add a transaction to this atomic group.
+*
+* An error will be thrown if the composer's status is not BUILDING, or if adding this transaction
+* causes the current group to exceed MAX_GROUP_SIZE.
+ */
+func (atc *AtomicTransactionComposer) AddTransaction(txnAndSigner TransactionWithSigner) error {
+	if atc.status != BUILDING {
+		return errors.New("STATUS MUST BE BUILDING IN ORDER TO ADD TRANSACTION")
+	}
+
+	if atc.Count() == MAX_GROUP_SIZE {
+		return fmt.Errorf("REACHED MAX_GROUP_SIZE: %d", MAX_GROUP_SIZE)
+	}
+
+	atc.transactions = append(atc.transactions, txnAndSigner)
+	return nil
+}
+
+/**
+* Add a smart contract method call to this atomic group.
+*
+* An error will be thrown if the composer's status is not BUILDING, if adding this transaction
+* causes the current group to exceed MAX_GROUP_SIZE, or if the provided arguments are invalid
+* for the given method.
+ */
+func (atc *AtomicTransactionComposer) AddMethodCall(
+	/** The ID of the smart contract to call */
+	appID uint64,
+	/** The method to call on the smart contract */
+	method Method,
+	/** The arguments to include in the method call. If omitted, no arguments will be passed to the method. */
+	methodArgs []MethodArgument,
+	/** The address of the sender of this application call */
+	sender types.Address,
+	/** Transactions params to use for this application call */
+	suggestedParams types.SuggestedParams,
+	/** The note value for this application call */
+	note []byte,
+	/** The lease value for this application call */
+	lease [32]byte,
+	/** If provided, the address that the sender will be rekeyed to at the conclusion of this application call */
+	rekeyTo types.Address,
+	/** A transaction signer that can authorize this application call from sender */
+	signer TransactionSigner) error {
+
+	if atc.status != BUILDING {
+		return errors.New("STATUS MUST BE BUILDING IN ORDER TO ADD TRANSACTION")
+	}
+
+	selectorValue := method.GetSelector()
+	encodedAbiArgs := [][]byte{selectorValue}
+
+	// insert selector as first param
+	var abiArgs []ABIValue
+	for _, methodArg := range methodArgs {
+		switch v := methodArg.(type) {
+		case ABIValue:
+			abiArgs = append(abiArgs, v)
+		case TransactionWithSigner:
+			err := atc.AddTransaction(v)
+			if err != nil {
+				return err
+			}
+		default:
+			return errors.New("MethodArg must be either ABIValue or TransactionSigner")
+		}
+	}
+
+	if atc.Count() == MAX_GROUP_SIZE {
+		return fmt.Errorf("REACHED MAX_GROUP_SIZE: %d", MAX_GROUP_SIZE)
+	}
+
+	for _, abiArg := range abiArgs {
+		encodedArg, err := abiArg.AbiType.Encode(abiArg.RawValue)
+		if err != nil {
+			return err
+		}
+
+		encodedAbiArgs = append(encodedAbiArgs, encodedArg)
+	}
+
+	tx, err := MakeApplicationNoOpTx(
```

Comment:
I initially and naively thought this was okay (my bad!), but now I do think we should follow the specs and use a `MakeApplicationCallTx` and accept the `OnComplete` params. Open to counterarguments.